### PR TITLE
Migrate KVCache to NNX

### PR DESCRIPTION
**Pull request by bvandermoon**
# Description

Use NNX for the KVCache module.

* Change module to NNX
* Initialize caches in __init__ instead of __call__
* Created properties for the existing cache tuples. This is because the paths were not being parsed properly in maxengine `tree_map_with_path` with NNX
* Used self variables for shapes instead of the shapes passed into __call__ wherever possible
* Create bridge to Linen function and use it where the KVCache module is invoked

# Tests

* Ran decode on v6e-8 devbox for `llama3.1-8B` (standard KVCache) and `deepseek3-test` (MlaKVCache). Got output for both.

```
python3 -m MaxText.decode MaxText/configs/base.yml \
    model_name=deepseek3-test \
    tokenizer_path=assets/tokenizer_llama3.tiktoken \
    tokenizer_type=tiktoken \
    scan_layers=false \
    per_device_batch_size=1 \
    ici_fsdp_parallelism=1 \
    ici_autoregressive_parallelism=-1 \
    max_prefill_predict_length=128 \
    max_target_length=256 \
    prompt="I love to" \
    attention=dot_product \
    mla_naive_kvcache=False
```
```
python3 -m MaxText.decode MaxText/configs/base.yml \
    model_name=llama3.1-8b \
    tokenizer_path=assets/tokenizer_llama3.tiktoken \
    tokenizer_type=tiktoken \
    scan_layers=false \
    per_device_batch_size=1 \
    ici_fsdp_parallelism=1 \
    ici_autoregressive_parallelism=-1 \
    max_prefill_predict_length=128 \
    max_target_length=256 \
    prompt="I love to" \
attention=dot_product
```

* kvcache_test.py: `python3 -m pytest MaxText/tests/inference/kvcache_test.py`

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/tests/inference/kvcache_test.py [hunk 1]
```diff
@@ -107,25 +105,22 @@ def test_update_kv_cache(self):
...
     key_rope_1 = jnp.ones((self.batchsize, 1, 1, self.k_rope_head_dim), dtype=self.dtype) * 0.05
-    _, new_vars = test_module.apply(
-        new_vars,
+    _ = test_module(
         low_rank_main_1,
         key_rope_1,
         decoder_segment_ids,
         model_mode,
-        rngs={"params": random.PRNGKey(0)},
-        mutable=True,
```
Comment (bvandermoon) [inline MaxText/tests/inference/kvcache_test.py:117]: @cgarciae looks like we need to continue passing `model_mode` into the __call__ functions based on this unit test. The `MlaKVCache` is initialized once but called with different `model_mode` values

@mitalisi @patemotter @jrplatin let us know if there is a way to set `model_mode` as a self variable in __init__. It would help when we remove the `KVCache` Linen bridges

File: MaxText/inference_microbenchmark.py [hunk 2]
```diff
@@ -155,6 +157,12 @@ def ar_benchmark_loop(config, engine_generate, params, decode_state, iters, prof
   for _ in range(iters):
     rng, rng_generate = jax.random.split(rng)
     decode_state, _ = engine_generate(params, decode_state, rng_generate)
+    decode_state = jax.tree.map(
```
Comment (cgarciae) [inline MaxText/inference_microbenchmark.py]: I think there's a helper that does exactly this, could make it a bit cleaner

**Review by cgarciae — APPROVED** [review]

File: MaxText/inference_microbenchmark.py (hunk 2 above)
Comment (bvandermoon) [reply MaxText/inference_microbenchmark.py]: Updated, thanks @cgarciae

File: MaxText/inference/kvcache.py [hunk 3]
```diff
@@ -149,33 +153,147 @@ def einsum_fn_with_rhs_qtensor_and_dequant(self, value):
...
+          CACHE_SCALE_KV,
+    ),
+    prefill_cache_axis_order: AxisIdxes = (1, 2, 0, 3),
+    ar_cache_axis_order: AxisIdxes = (1, 2, 0, 3),
+    key_axis_order: AxisIdxes = (2, 0, 1, 3),
+    use_chunked_prefill: bool = False,
+    model_mode: str = MODEL_MODE_PREFILL,
+    name: str | None = None,
+):
+  """Initializes the KVCache module and returns it as a Linen module."""
```
Comment (jrplatin) [inline MaxText/inference/kvcache.py]: Can you add a docstring here?

File: MaxText/inference/kvcache.py [hunk 4]
```diff
@@ -149,33 +153,147 @@ def einsum_fn_with_rhs_qtensor_and_dequant(self, value):
...
+      prefill_cache_axis_order: AxisIdxes = (1, 2, 0, 3),
+      ar_cache_axis_order: AxisIdxes = (1, 2, 0, 3),
+      key_axis_order: AxisIdxes = (2, 0, 1, 3),
+      use_chunked_prefill: bool = False,
+      model_mode: str = MODEL_MODE_PREFILL,
+      *,
+      # Not used in KVCache but passed in by nnx_wrappers.to_linen.
+      # TODO: Remove when bridge no longer needed
+      rngs: nnx.Rngs = None,
+  ):
```
Comment (jrplatin) [inline MaxText/inference/kvcache.py:273]: Same here -- would be great to have some description of these vars

File: MaxText/inference/kvcache.py [hunk 5]
```diff
@@ -373,27 +463,20 @@ def kv_cache_chunked_prefill(
     """
 
     assert not self.kv_quant, "Not support kv_quant now."
-    batch, key_seq_len, key_heads, key_head_size = key.shape
-    batch, value_seq_len, value_heads, value_head_size = value.shape
     if decoder_segment_ids is not None:
-      batch, segment_id_seq_len = decoder_segment_ids.shape
-      assert key_seq_len == segment_id_seq_len, f"{key_seq_len=}, {segment_id_seq_len=} should match."
+      self.batch, segment_id_seq_len = decoder_segment_ids.shape
+      assert self.key_seq_len == segment_id_seq_len, f"{self.key_seq_len=}, {segment_id_seq_len=} should match."
```
Comment (jrplatin) [inline MaxText/inference/kvcache.py:520]: nit: I think Ran and a few others mentioned moving away from `assert` to using `raise` instead?

File: MaxText/layers/attentions.py [hunk 6]
```diff
@@ -1649,19 +1649,31 @@ def apply_rotary_embedding(self, inputs: Array, name: str, inputs_positions: Opt
 
   def update_kv_caches(self, key, value, decoder_segment_ids, model_mode, previous_chunk):
     """Updates the KV caches for prefill and autoregressive modes."""
```
Comment (jrplatin) [inline MaxText/layers/attentions.py]: Would be great to have docstrings here as well

**Review by jrplatin — APPROVED** [review]
LGTM -- didn't review the MLA work closely but assuming your Llama/DeepSeek E2E tests pass and match current throughput/accuracy numbers, then I don't see an issue

File: MaxText/inference/kvcache.py (hunk 4 above)
Comment (bvandermoon) [reply MaxText/inference/kvcache.py:273]: Added, thanks Jacob

File: MaxText/inference/kvcache.py (hunk 5 above)
Comment (bvandermoon) [reply MaxText/inference/kvcache.py:520]: Makes sense. We are trying to avoid functional changes due to the nature of the NNX migration though. This would significantly increase the scope and also make it harder to test correctness. Let me know if you have any concerns here

File: MaxText/inference/kvcache.py (hunk 3 above)
Comment (bvandermoon) [reply MaxText/inference/kvcache.py]: Added, thanks Jacob

File: MaxText/layers/attentions.py (hunk 6 above)
Comment (bvandermoon) [reply MaxText/layers/attentions.py]: Added, thanks Jacob

**Review by richjames0 — APPROVED** [review]
lgtm
